File: 1/js/src/BST.ts
class BSTNode {
	data: number;
	left: BSTNode | null;
	right: BSTNode | null;

	constructor(data: number) {
		this.data = data;
		this.left = this.right = null;
	}
}

class BST {
	root: BSTNode | null;

	constructor() {
		this.root = null;
	}

	insertOne(data: number) {
		const newNode = new BSTNode(data);	

		let parentNode: BSTNode | null = null;
		let currentNode: BSTNode | null = this.root; 
		let isLeftNode  = false;	

		while(currentNode != null) {
			parentNode = currentNode;
			if(data < currentNode.data) {
				isLeftNode = true;	
				currentNode = currentNode.left;
			} else {
				isLeftNode = false;
				currentNode = currentNode.right;
			}
		}
		if(parentNode == null) {
			this.root = newNode;
		} else if(isLeftNode) {
			parentNode.left = newNode;
		}	else {
			parentNode.right = newNode;
		}
	}

	insertMany(dataArr: number[]) {
		for(const data of dataArr) {
			this.insertOne(data);
		}
	}
}

export {BSTNode, BST};

File: 1/java/BST.java
public class BST {
	private Node root = null;	

	public class Node {
		public int value;
		public Node left = null;
		public Node right = null;

		Node(int value) {
			this.value = value;
		}
	}

	public void insert(int value) {
	
		Node prev = null;
		Node temp = root;

		while(temp != null) {
			prev = temp;

			if(value < temp.value) {
				temp = temp.left;
			} else {
				temp = temp.right;
			}
		}

		if (prev == null) {
			root = new Node(value);
		} else if(value < prev.value) {
			prev.left = new Node(value);
		} else {
			prev.right = new Node(value);
		}
	}
	
	public void insert(int arr[]) {
		for(int i: arr) {
			this.insert(i);
		}
	}

	public Node getRoot() {
		return root;
	}
	
}
